modules/codegen/src/main/scala/com/twilio/guardrail/generators/EndpointsClientGenerator.scala

```diff
@@ -0,0 +1,462 @@
+            if (tracing) { (Some(q"tracerHeader"), Some(q"Map.empty")) } else { (None, None) }
+
+          val responseCompanionTerm = Term.Name(s"${methodName.capitalize}Response")
+          val responseCompanionType = Type.Name(s"${methodName.capitalize}Response")
+
+          val cases = responses.value.map { resp =>
+            val responseTerm = Term.Name(s"${resp.statusCodeName.value}")
+            resp.value.fold[Case](
+              p"case ${Lit.Int(resp.statusCode)} => Right($responseCompanionTerm.$responseTerm)"
+            ) {
+              case (tpe, _) =>
+                p"""case ${Lit.Int(resp.statusCode)} =>
+                  parser.parse(xhr.responseText).flatMap(CirceDecoder[${tpe}].decodeJson _).map($responseCompanionTerm.$responseTerm.apply _)
+                """
+            }
+          } :+ p"case _ => Left(new UnknownStatusException(xhr))"
+
+          val responseTypeRef = Type.Name(s"${methodName.capitalize}Response")
+
+          val methodParameters: List[List[Term.Param]] = List(
+            Some(
+              tracingArgsPre.map(_.param) ++ pathArgs.map(_.param) ++ qsArgs
+                .map(_.param) ++ formArgs.map(_.param) ++ bodyArgs
+                .map(_.param) ++ headerArgs.map(_.param) ++ tracingArgsPost
+                .map(_.param)
+            ),
+            implicitParams
+          ).flatten
+
+          val endpointCallArgs = {
+            def adaptNel: NonEmptyList[Term] => Term = {
+              case NonEmptyList(a, Nil)     => a
+              case NonEmptyList(a, x :: xs) => emulateTupler(a, NonEmptyList(x, xs))
+            }
+            @scala.annotation.tailrec
+            def emulateTupler(base: Term, xs: NonEmptyList[Term]): Term.Tuple = xs match {
+              case NonEmptyList(x, Nil) =>
+                base match {
+                  case q"($a, $b)" => q"($a, $b, $x)"
+                  case as          => q"($as, $x)"
+                }
+              case NonEmptyList(x, y :: Nil) =>
+                base match {
+                  case q"($a, $b)" => q"($a, $b, $x, $y)"
+                  case as          => q"($as, $x, $y)"
+                }
+              case NonEmptyList(x, y :: rest) =>
+                emulateTupler(base match {
+                  case q"($a, $b)" => q"($a, $b, $x)"
+                  case as          => q"($as, $x)"
+                }, NonEmptyList(y, rest))
+            }
+
+            def hackyFoldLimitedTupleTree: NonEmptyList[Term] => NonEmptyList[Term] = {
+              case NonEmptyList(x, xs) =>
+                // FIXME: There's a hard limit in endpoints for a maximum Tupler size of 3 elements.
+                // This results in ((((1, 2, 3), 4, 5), 6, 7)...)
+                // As a result, we've got a rather complex structure to replicate this,
+                // bounded by ${limit}. If this limit is ever raised, it'll be a breaking change, unless
+                // this value becomes customizable.
```

Comment:
That's... unfortunate.

Reply:
yes.
[Twitter: ![image](https://user-images.githubusercontent.com/278900/53044988-5aa82b80-3441-11e9-8368-4ef67fa2a7f3.png)](https://twitter.com/blast_hardchese/status/1087088858368360448)

Reply:
Oh man...